File: algorithm/31. Next Permutation/31.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int index = nums.size() - 1;
        while(nums[index] <= nums[index-1]) {
            --index;
        }
        if(index == 0) {
            sort(nums.begin(), nums.end());
            return ;
        }
        int exchangeIndex;
        for(int i = nums.size()-1; i >= index; i--){
            if(nums[i] > nums[index-1]){
                exchangeIndex = i;
                break;
            }
        }
        swap(nums[index-1],nums[exchangeIndex]);
        sort(nums.begin()+index, nums.end());    
    }
};

File: algorithm/31. Next Permutation/31.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        index = len(nums)-1
        while index>0 and nums[index]<=nums[index-1]:
            index-=1
        if index==0:
            nums.sort()
            return
        exchange=0
        for i in range(len(nums)-1,index-1,-1):
            if nums[i]>nums[index-1]:
                exchange=i
                break
        tem=nums[index-1]
        nums[index-1]=nums[exchange]
        nums[exchange]=tem
        nums[index:] = nums[:index-1:-1]
        return 
